File: Year_III/SEM_2/CT - Compilation Techniques/LAB/Project Python/libs/syntactical_analyzer.py
from typing import List, Tuple, Callable
from .token import *


class Parser:
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.current = 0

    def consume(self, token_code: TokenCode) -> bool:
        if self.tokens[self.current].code == token_code:
            print(f'Consumed {self.tokens[self.current]}')
            self.current += 1
            return True
        # print(f'Expected {token_code.name}')
        return False

    def apply(self, rule: Callable[['Parser'], bool], *args) -> bool:
        current = self.current

        # print(f'Applying rule {rule.__name__}')
        if rule(self, *args):
            return True

        self.current = current
        return False


# variable types
def rule_type(parser: Parser) -> bool:
    if parser.consume(TokenCode.INT):
        return True
    if parser.consume(TokenCode.DOUBLE):
        return True
    if parser.consume(TokenCode.CHAR):
        return True
    if parser.consume(TokenCode.STRUCT):
        if not parser.consume(TokenCode.ID):
            return False
        return True

    return False


def rule_type_name(parser: Parser) -> bool:
    if not parser.apply(rule_type):
        return False
    parser.apply(rule_array)

    return True

# argument list
def rule_argument_list(parser: Parser) -> bool:
    if not parser.apply(rule_expression):
        return False
    while parser.consume(TokenCode.COMMA):
        if not parser.apply(rule_expression):
            return False

    return True

# array declaration
def rule_array(parser: Parser) -> bool:
    if not parser.consume(TokenCode.LBRACKET):
        return False
    parser.apply(rule_expression)
    if not parser.consume(TokenCode.RBRACKET):
        return False

    return True

# expressions
def rule_expression(parser: Parser) -> bool:
    if not parser.apply(rule_expression_assign):
        return False

    return True


def rule_expression_assign(parser: Parser) -> bool:
    if not parser.apply(rule_expression_or):
        return False
    if parser.consume(TokenCode.ASSIGN):
        if not parser.apply(rule_expression_assign):
            return False

    return True


def rule_expression_or(parser: Parser) -> bool:
    if not parser.apply(rule_expression_and):
        return False
    while parser.consume(TokenCode.OR):
        if not parser.apply(rule_expression_and):
            return False

    return True


def rule_expression_and(parser: Parser) -> bool:
    if not parser.apply(rule_expression_equality):
        return False
    while parser.consume(TokenCode.AND):
        if not parser.apply(rule_expression_equality):
            return False

    return True


def rule_expression_equality(parser: Parser) -> bool:
    if not parser.apply(rule_expression_relational):
        return False
    while parser.consume(TokenCode.EQ) or parser.consume(TokenCode.NE):
        if not parser.apply(rule_expression_relational):
            return False

    return True


def rule_expression_relational(parser: Parser) -> bool:
    if not parser.apply(rule_expression_add_sub):
        return False
    while parser.consume(TokenCode.LT) or parser.consume(TokenCode.LE) or parser.consume(TokenCode.GT) or parser.consume(TokenCode.GE):
        if not parser.apply(rule_expression_add_sub):
            return False

    return True


def rule_expression_add_sub(parser: Parser) -> bool:
    if not parser.apply(rule_expression_mul_div):
        return False
    while parser.consume(TokenCode.ADD) or parser.consume(TokenCode.SUB):
        if not parser.apply(rule_expression_mul_div):
            return False

    return True


def rule_expression_mul_div(parser: Parser) -> bool:
    if not parser.apply(rule_expression_cast):
        return False
    while parser.consume(TokenCode.MUL) or parser.consume(TokenCode.DIV):
        if not parser.apply(rule_expression_cast):
            return False

    return True


def rule_expression_cast(parser: Parser) -> bool:
    if parser.consume(TokenCode.LPAREN):
        if not parser.apply(rule_type_name):
            return False
        if not parser.consume(TokenCode.RPAREN):
            return False
        if not parser.apply(rule_expression_cast):
            return False

        return True
    if not parser.apply(rule_expression_unary):
        return False

    return True


def rule_expression_unary(parser: Parser) -> bool:
    if parser.consume(TokenCode.ADD) or parser.consume(TokenCode.SUB):
        if not parser.apply(rule_expression_unary):
            return False

        return True
    if not parser.apply(rule_expression_postfix):
        return False

    return True


def rule_expression_postfix(parser: Parser) -> bool:
    if not parser.apply(rule_expression_primary):
        return False
    while (1):
        if parser.consume(TokenCode.LBRACKET):
            if not parser.apply(rule_expression):
                return False
            if not parser.consume(TokenCode.RBRACKET):
                return False
        elif parser.consume(TokenCode.DOT):
            if not parser.consume(TokenCode.ID):
                return False
        else:
            break

    return True


def rule_expression_primary(parser: Parser) -> bool:
    if parser.consume(TokenCode.ID):
        if parser.consume(TokenCode.LPAREN):
            parser.apply(rule_argument_list)
            if not parser.consume(TokenCode.RPAREN):
                return False

            return True
        return True
    if parser.consume(TokenCode.CT_INT):
        return True
    if parser.consume(TokenCode.CT_REAL):
        return True
    if parser.consume(TokenCode.CT_CHAR):
        return True
    if parser.consume(TokenCode.CT_STRING):
        return True
    if parser.consume(TokenCode.LPAREN):
        if not parser.apply(rule_expression):
            return False
        if not parser.consume(TokenCode.RPAREN):
            return False

        return True

    return False

# statements
def rule_statement_list(parser: Parser) -> bool:
    if not parser.consume(TokenCode.LBRACE):
        return False
    while parser.apply(rule_declare_var) or parser.apply(rule_statement):
        pass
    if not parser.consume(TokenCode.RBRACE):
        return False

    return True


def rule_statement_else(parser: Parser) -> bool:
    if parser.consume(TokenCode.ELSE):
        if not parser.apply(rule_statement):
            return False

    return True


def rule_statement_if(parser: Parser) -> bool:
    if not parser.consume(TokenCode.IF):
        return False
    if not parser.consume(TokenCode.LPAREN):
        return False
    if not parser.apply(rule_expression):
        return False
    if not parser.consume(TokenCode.RPAREN):
        return False
    if not parser.apply(rule_statement):
        return False
    parser.apply(rule_statement_else)

    return True


def rule_statement_while(parser: Parser) -> bool:
    if not parser.consume(TokenCode.WHILE):
        return False
    if not parser.consume(TokenCode.LPAREN):
        return False
    if not parser.apply(rule_expression):
        return False
    if not parser.consume(TokenCode.RPAREN):
        return False
    if not parser.apply(rule_statement):
        return False

    return True


def rule_statement_for(parser: Parser) -> bool:
    if not parser.consume(TokenCode.FOR):
        return False
    if not parser.consume(TokenCode.LPAREN):
        return False
    parser.apply(rule_expression)
    if not parser.consume(TokenCode.SEMICOLON):
        return False
    parser.apply(rule_expression)
    if not parser.consume(TokenCode.SEMICOLON):
        return False
    parser.apply(rule_expression)
    if not parser.consume(TokenCode.RPAREN):
        return False
    if not parser.apply(rule_statement):
        return False

    return True


def rule_statement_break(parser: Parser) -> bool:
    if not parser.consume(TokenCode.BREAK):
        return False
    if not parser.consume(TokenCode.SEMICOLON):
        return False

    return True


def rule_statement_return(parser: Parser) -> bool:
    if not parser.consume(TokenCode.RETURN):
        return False
    parser.apply(rule_expression)
    if not parser.consume(TokenCode.SEMICOLON):
        return False

    return True


def rule_statement_expression(parser: Parser) -> bool:
    parser.apply(rule_expression)
    if not parser.consume(TokenCode.SEMICOLON):
        return False

    return True


def rule_statement(parser: Parser) -> bool:
    if parser.apply(rule_statement_list):
        return True
    if parser.apply(rule_statement_if):
        return True
    if parser.apply(rule_statement_while):
        return True
    if parser.apply(rule_statement_for):
        return True
    if parser.apply(rule_statement_break):
        return True
    if parser.apply(rule_statement_return):
        return True
    if parser.apply(rule_statement_expression):
        return True

    return False

# declaration
def rule_declare_var(parser: Parser) -> bool:
    if not parser.apply(rule_type):
        return False
    if not parser.consume(TokenCode.ID):
        return False
    parser.apply(rule_array)
    while parser.consume(TokenCode.COMMA):
        if not parser.consume(TokenCode.ID):
            return False
        parser.apply(rule_array)
    if not parser.consume(TokenCode.SEMICOLON):
        return False

    return True


def rule_declare_struct(parser: Parser) -> bool:
    if not parser.consume(TokenCode.STRUCT):
        return False
    if not parser.consume(TokenCode.ID):
        return False
    if not parser.consume(TokenCode.LBRACE):
        return False
    while parser.apply(rule_declare_var):
        pass
    if not parser.consume(TokenCode.RBRACE):
        return False
    if not parser.consume(TokenCode.SEMICOLON):
        return False

    return True


def rule_func_param(parser: Parser) -> bool:
    if not parser.apply(rule_type):
        return False
    if not parser.consume(TokenCode.ID):
        return False
    parser.apply(rule_array)

    return True


def rule_func_params(parser: Parser) -> bool:
    if not parser.apply(rule_func_param):
        return False
    while parser.consume(TokenCode.COMMA):
        if not parser.apply(rule_func_param):
            return False

    return True


def rule_declare_func(parser: Parser) -> bool:
    if not parser.apply(rule_type) and not parser.consume(TokenCode.VOID):
        return False
    if not parser.consume(TokenCode.ID):
        return False
    if not parser.consume(TokenCode.LPAREN):
        return False
    parser.apply(rule_func_params)
    if not parser.consume(TokenCode.RPAREN):
        return False
    if not parser.apply(rule_statement_list):
        return False

    return True

# code
def rule_code(parser: Parser) -> bool:
    while parser.apply(rule_declare_struct) or parser.apply(rule_declare_func) or parser.apply(rule_declare_var):
        pass
    if not parser.consume(TokenCode.END):
        return False

    return True


def analyze_syntactically(tokens: List[Token]):
    parser = Parser(tokens)
    parser.apply(rule_code)

File: Year_III/SEM_2/CT - Compilation Techniques/LAB/Project C/lib/syntactical_analyzer.c
#include "syntactical_analyzer.h"

token_t *consumed_token;
token_t *current_token;

int consumeToken(int type) {
    if (current_token->type == type) {
        consumed_token = current_token;
        current_token = current_token->next;

        return 1;
    }
    return 0;
}

/// RULES
// PROTOTYPES
int ruleCode();
int ruleStatementList();
int ruleStatementElse();
int ruleStatementIf();
int ruleStatementWhile();
int ruleStatementFor();
int ruleStatementBreak();
int ruleStatementReturn();
int ruleStatementExpression();
int ruleStatement();
int ruleType();
int ruleTypeName();
int ruleArgumentList();
int ruleExpression();
int ruleExpressionAssign();
int ruleExpressionOr();
int ruleExpressionAnd();
int ruleExpressionEquality();
int ruleExpressionRelational();
int ruleExpressionAddSub();
int ruleExpressionMulDiv();
int ruleExpressionCast();
int ruleExpressionUnary();
int ruleExpressionPostfix();
int ruleExpressionPrimary();
int ruleArray();
int ruleDeclareVar();
int ruleDeclareStruct();
int ruleFuncParam();
int ruleFuncParams();
int ruleDeclareFunc();

// variable types
int ruleType() {
    if (consumeToken(INT)) return 1;
    if (consumeToken(DOUBLE)) return 1;
    if (consumeToken(CHAR)) return 1;
    if (consumeToken(STRUCT)) {
        if (!consumeToken(ID)) return 0;
        return 1;
    }

    return 0;
}

int ruleTypeName() {
    if (!ruleType()) return 0;
    ruleArray();

    return 1;
}

// argument list
int ruleArgumentList() {
    if (!ruleExpression()) return 0;
    while (consumeToken(COMMA)) {
        if (!ruleExpression()) return 0;
    }

    return 1;
}

// array declaration
int ruleArray() {
    if (!consumeToken(LBRACKET)) return 0;
    ruleExpression();
    if (!consumeToken(RBRACKET)) return 0;

    return 1;
}

// expressions
int ruleExpression() {
    if (!ruleExpressionAssign()) return 0;

    return 1;
}

int ruleExpressionAssign() {
    if (!ruleExpressionOr()) return 0;
    if (consumeToken(ASSIGN)) {
        if (!ruleExpressionAssign()) return 0;
    }

    return 1;
}

int ruleExpressionOr() {
    if (!ruleExpressionAnd()) return 0;
    while (consumeToken(OR)) {
        if (!ruleExpressionAnd()) return 0;
    }

    return 1;
}

int ruleExpressionAnd() {
    if (!ruleExpressionEquality()) return 0;
    while (consumeToken(AND)) {
        if (!ruleExpressionEquality()) return 0;
    }

    return 1;
}

int ruleExpressionEquality() {
    if (!ruleExpressionRelational()) return 0;
    while (consumeToken(EQ) || consumeToken(NE)) {
        if (!ruleExpressionRelational()) return 0;
    }

    return 1;
}

int ruleExpressionRelational() {
    if (!ruleExpressionAddSub()) return 0;
    while (consumeToken(LT) || consumeToken(LE) || consumeToken(GT) || consumeToken(GE)) {
        if (!ruleExpressionAddSub()) return 0;
    }

    return 1;
}

int ruleExpressionAddSub() {
    if (!ruleExpressionMulDiv()) return 0;
    while (consumeToken(ADD) || consumeToken(SUB)) {
        if (!ruleExpressionMulDiv()) return 0;
    }

    return 1;
}

int ruleExpressionMulDiv() {
    if (!ruleExpressionCast()) return 0;
    while (consumeToken(MUL) || consumeToken(DIV)) {
        if (!ruleExpressionCast()) return 0;
    }

    return 1;
}

int ruleExpressionCast() {
    if (consumeToken(LPAREN)) {
        if (!ruleTypeName()) return 0;
        if (!consumeToken(RPAREN)) return 0;
        if (!ruleExpressionCast()) return 0;

        return 1;
    }
    if (!ruleExpressionUnary()) return 0;

    return 1;
}

int ruleExpressionUnary() {
    if (consumeToken(ADD) || consumeToken(SUB)) {
        if (!ruleExpressionUnary()) return 0;

        return 1;
    }
    if (!ruleExpressionPostfix()) return 0;

    return 1;
}

int ruleExpressionPostfix() {
    if (!ruleExpressionPrimary()) return 0;
    while (1) {
        if (consumeToken(LBRACKET)) {
            if (!ruleExpression()) return 0;
            if (!consumeToken(RBRACKET)) return 0;
        } else if (consumeToken(DOT)) {
            if (!consumeToken(ID)) return 0;
        } else {
            break;
        }
    }

    return 1;
}

int ruleExpressionPrimary() {
    if (consumeToken(ID)) {
        if (consumeToken(LPAREN)) {
            ruleArgumentList();
            if (!consumeToken(RPAREN)) return 0;

            return 1;
        }
        return 1;
    }
    if (consumeToken(CT_INT)) return 1;
    if (consumeToken(CT_REAL)) return 1;
    if (consumeToken(CT_CHAR)) return 1;
    if (consumeToken(CT_STRING)) return 1;
    if (consumeToken(LPAREN)) {
        if (!ruleExpression()) return 0;
        if (!consumeToken(RPAREN)) return 0;

        return 1;
    }

    return 0;
}

// statements
int ruleStatementList() {
    if (!consumeToken(LBRACE)) return 0;
    while (ruleDeclareVar() || ruleStatement());
    if (!consumeToken(RBRACE)) return 0;

    return 1;
}

int ruleStatementElse() {
    if (consumeToken(ELSE)) {
        if (!ruleStatement()) return 0;
    }

    return 1;
}

int ruleStatementIf() {
    if (!consumeToken(IF)) return 0;
    if (!consumeToken(LPAREN)) return 0;
    if (!ruleExpression()) return 0;
    if (!consumeToken(RPAREN)) return 0;
    if (!ruleStatement()) return 0;
    ruleStatementElse();

    return 1;
}

int ruleStatementWhile() {
    if (!consumeToken(WHILE)) return 0;
    if (!consumeToken(LPAREN)) return 0;
    if (!ruleExpression()) return 0;
    if (!consumeToken(RPAREN)) return 0;
    if (!ruleStatement()) return 0;

    return 1;
}

int ruleStatementFor() {
    if (!consumeToken(FOR)) return 0;
    if (!consumeToken(LPAREN)) return 0;
    ruleExpression();
    if (!consumeToken(SEMICOLON)) return 0;
    ruleExpression();
    if (!consumeToken(SEMICOLON)) return 0;
    ruleExpression();
    if (!consumeToken(RPAREN)) return 0;
    if (!ruleStatement()) return 0;

    return 1;
}

int ruleStatementBreak() {
    if (!consumeToken(BREAK)) return 0;
    if (!consumeToken(SEMICOLON)) return 0;

    return 1;
}

int ruleStatementReturn() {
    if (!consumeToken(RETURN)) return 0;
    ruleExpression();
    if (!consumeToken(SEMICOLON)) return 0;

    return 1;
}

int ruleStatementExpression() {
    ruleExpression();
    if (!consumeToken(SEMICOLON)) return 0;

    return 1;
}

int ruleStatement() {
    if (ruleStatementList()) return 1;
    if (ruleStatementIf()) return 1;
    if (ruleStatementWhile()) return 1;
    if (ruleStatementFor()) return 1;
    if (ruleStatementBreak()) return 1;
    if (ruleStatementReturn()) return 1;
    if (ruleStatementExpression()) return 1;

    return 0;
}

// declaration
int ruleDeclareVar() {
    if (!ruleType()) return 0;
    if (!consumeToken(ID)) return 0;
    ruleArray();
    while (consumeToken(COMMA)) {
        if (!consumeToken(ID)) return 0;
        ruleArray();
    }
    if (!consumeToken(SEMICOLON)) return 0;

    return 1;
}

int ruleDeclareStruct() {
    if (!consumeToken(STRUCT)) return 0;
    if (!consumeToken(ID)) return 0;
    if (!consumeToken(LBRACE)) return 0;
    while (ruleDeclareVar()) {
    }
    if (!consumeToken(RBRACE)) return 0;
    if (!consumeToken(SEMICOLON)) return 0;

    return 1;
}

int ruleFuncParam() {
    if (!ruleType()) return 0;
    if (!consumeToken(ID)) return 0;
    ruleArray();

    return 1;
}

int ruleFuncParams() {
    if (!ruleFuncParam()) return 0;
    while (consumeToken(COMMA)) {
        if (!ruleFuncParam()) return 0;
    }

    return 1;
}

int ruleDeclareFunc() {
    if (!ruleType() && !consumeToken(VOID)) return 0;
    if (!consumeToken(ID)) return 0;
    if (!consumeToken(LPAREN)) {
        return 0;    
    }
    ruleFuncParams();
    if (!consumeToken(RPAREN)) return 0;
    if (!ruleStatementList()) return 0;

    return 1;
}

// code
int ruleCode() {
    while (ruleDeclareStruct() || ruleDeclareFunc() || ruleDeclareVar()) {
    }
    if (!consumeToken(END)) return 0;

    return 1;
}